### PR TITLE
[NPU] Groupwise

### PR DESCRIPTION
**Pull request by cyita**
## Description

TODOs:
- [x] Remove redundant code
- [ ] ~Update readme~ -> next pr
- [ ] ~Support Llama~ -> next pr

### 2. User API changes
Add quantization_group_size=0/32/64/128 in from_pretrained
```
model = AutoModelForCausalLM.from_pretrained(model_path, load_in_low_bit=low_bit, trust_remote_code=True, torch_dtype=torch.float16,
                                                     optimize_model=optimize_model, max_output_len=max_output_len, max_prompt_len=int(in_out_len[0]), transpose_value_cache=transpose_value_cache,
                                                     use_cache=True, attn_implementation="eager", quantization_group_size=quantization_group_size).eval()
```

### 4. How to test?
- [ ] N/A
- [ ] Unit test: Please manually trigger the PR Validation [here](https://github.com/intel-analytics/ipex-llm-workflow/actions/workflows/llm-PR-validation.yml) by inputting the PR number (e.g., `1234`). And paste your action link here once it has been successfully finished.
- [ ] Application test
- [ ] Document test
- [ ] ...

### 5. New dependencies

- [ ] New Python dependencies
       - Dependency1 
       - Dependency2
       - ...
- [ ] New Java/Scala dependencies and their license
       - Dependency1 and license1
       - Dependency2 and license2
       - ...

### CODE REVIEW

File: python/llm/src/ipex_llm/transformers/npu_models/convert_mp.py [hunk 1]
```diff
@@ -86,17 +88,43 @@ def optimize_llm_pre(model: torch.nn.Module, qtype, mixed_precision):
...
+                if quantization_group_size == 0:
+                    split_num = 14
+                else:
+                    split_num = 28
+                    use_split = True
+            else:
+                split_num = 1
             new_lm_head = SlicedLMHead(model.lm_head.weight, split_num=split_num,
-                                       bias=model.lm_head.bias)
+                                       bias=model.lm_head.bias, use_split=use_split)
```
Comment (jason-dai) [inline python/llm/src/ipex_llm/transformers/npu_models/convert_mp.py:127]: How about:
```python
if quantization_group_size != 0
            split_num = model.config.hidden_size // quantization_group_size
            new_lm_head = SlicedLMHead(model.lm_head.weight, split_num=split_num,
                                       bias=model.lm_head.bias)
                                       bias=model.lm_head.bias, use_split=true)
else
           is_split = (not mixed_precision) and qtype == "sym_int4_rtn"
           split_num = 14 if is_split else 1
           new_lm_head = SlicedLMHead(model.lm_head.weight, split_num=split_num,
                                       bias=model.lm_head.bias)
```

File: python/llm/src/ipex_llm/transformers/npu_models/convert_mp.py [hunk 2]
```diff
@@ -132,6 +160,7 @@ def optimize_llm(
     inter_pp=None,
     intra_pp=None,
     transpose_value_cache=True,
+    is_groupwise_quant=False
```
Comment (jason-dai) [inline python/llm/src/ipex_llm/transformers/npu_models/convert_mp.py:163]: maybe use `group_size = 0` instead?

File: python/llm/src/ipex_llm/transformers/npu_models/linear.py [hunk 3]
```diff
@@ -130,6 +130,7 @@ def __init__(
         weight: torch.Tensor,
         scale: torch.Tensor,
         bias: Optional[torch.Tensor] = None,
+        is_groupwise_quant: bool = False,
```
Comment (jason-dai) [inline python/llm/src/ipex_llm/transformers/npu_models/linear.py:133]: maybe use `group_size = 0` instead?

File: python/llm/src/ipex_llm/transformers/npu_models/lm_head.py [hunk 4]
```diff
@@ -51,26 +53,45 @@ def __init__(
...
+                                           wt_dtype=dtype)
+                if i == 0:
+                    res = linear_slice
+                else:
+                    res += linear_slice
 
         print("start compiling lm_head")
         self.compile()
+        qwen_size = "7b" if self.inC == 3584 else "1.5b"
+        xml_path = f"qwen/qwen-{qwen_size}-npu-lmhead-{self.split_num}.xml"
```
Comment (jason-dai) [inline python/llm/src/ipex_llm/transformers/npu_models/lm_head.py:81]: why `qwen`?

**Comment by jason-dai** [comment]
maybe we can combine dq_split_linear & linear into one (i.e., a unified linear that automatically handles group_size/split_number) in future

File: python/llm/src/ipex_llm/transformers/npu_models/convert_mp.py (hunk 1 above)
Comment (cyita) [reply python/llm/src/ipex_llm/transformers/npu_models/convert_mp.py]: updated

File: python/llm/src/ipex_llm/transformers/npu_models/linear.py (hunk 3 above)
Comment (cyita) [reply python/llm/src/ipex_llm/transformers/npu_models/linear.py]: Updated

**Review by Oscilloscope98 — APPROVED** [review]
LGTM

We may deal with `quantization_group_size` for save & load model later

**Comment by cyita** [comment]
The performance of the loaded model in this PR is slower than `from_pretained` for an unknown reason, which will be fixed in the next PR.

**Pull request merged**